pkg/networkservice/common/heal/eventloop.go

```diff
@@ -66,69 +68,78 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
 	client, err := networkservice.NewMonitorConnectionClient(cc).MonitorConnections(eventLoopCtx, selector)
 	if err != nil {
 		eventLoopCancel()
-		return nil, errors.Wrap(err, "failed get MonitorConnections client")
+		return nil, nil, errors.Wrap(err, "failed get MonitorConnections client")
 	}
 
 	// get the initial state transfer and use it to detect whether we have a real connection or not
 	_, err = client.Recv()
 	if err != nil {
 		eventLoopCancel()
-		return nil, errors.Wrap(err, "failed to get the initial state transfer")
+		return nil, nil, errors.Wrap(err, "failed to get the initial state transfer")
 	}
 
 	logger := log.FromContext(ctx).WithField("heal", "eventLoop")
 	cev := &eventLoop{
-		heal:         heal,
-		eventLoopCtx: eventLoopCtx,
-		chainCtx:     ctx,
-		conn:         conn,
-		eventFactory: ev,
-		client:       newClientFilter(client, conn, logger),
-		logger:       logger,
+		heal:              heal,
+		eventLoopCtx:      eventLoopCtx,
+		eventLoopCancel:   eventLoopCancel,
+		chainCtx:          ctx,
+		conn:              conn,
+		eventFactory:      ev,
+		client:            newClientFilter(client, conn, logger),
+		logger:            logger,
+		monitorFinishedCh: make(chan bool, 1),
 	}
 
 	// Start the eventLoop
 	go cev.eventLoop()
-	return eventLoopCancel, nil
+	return eventLoopCancel, cev.monitorFinishedCh, nil
 }
 
-func (cev *eventLoop) monitorCtrlPlane() <-chan struct{} {
-	res := make(chan struct{}, 1)
-
-	go func() {
-		defer close(res)
+func (cev *eventLoop) eventLoop() {
+	needToHeal, reselect := cev.waitForEvents()
 
-		for {
-			eventIn, err := cev.client.Recv()
+	if !needToHeal {
```

Comment:
Probably name inversion could be more clear

```suggestion
	if healthy {
```

Reply:
What if needToHead=false and reselct=true?

Reply:
In the event monitoring there are three states:
1. Context was cancelled, `needToHeal == false`
2. Control plane is down, `needToHeal == true`, `reselect == false`
3. Data plane is down, `needToHeal == true`, `reselect == true`

Maybe we could add a special enum for these three states, but I think the code would only become more complex, because these 2 variables are checked and changed in different places.

Reply:
> Probably name inversion could be more clear

Fixed.
I modified it to use `canceled`.